File: openvc/src/main/java/io/github/_4drian3d/authmevelocity/velocity/AuthMeVelocityPlugin.java
/*
 * Copyright (C) 2025 AuthMeVelocity Contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package io.github._4drian3d.authmevelocity.velocity;

import static net.kyori.adventure.text.minimessage.MiniMessage.miniMessage;

import com.google.inject.Injector;
import com.velocitypowered.api.proxy.Player;
import com.velocitypowered.api.proxy.ProxyServer;
import com.velocitypowered.api.proxy.ServerConnection;
import com.velocitypowered.api.proxy.messages.ChannelIdentifier;
import com.velocitypowered.api.proxy.messages.LegacyChannelIdentifier;
import com.velocitypowered.api.proxy.messages.MinecraftChannelIdentifier;
import com.velocitypowered.api.proxy.server.RegisteredServer;
import fun.iiii.mixedlogin.config.MixedLoginConfig;
import io.github._4drian3d.authmevelocity.velocity.listener.Listener;
import io.github._4drian3d.authmevelocity.velocity.listener.connection.DisconnectListener;
import io.github._4drian3d.authmevelocity.velocity.listener.connection.InitialServerListener;
import io.github._4drian3d.authmevelocity.velocity.listener.connection.PostConnectListener;
import io.github._4drian3d.authmevelocity.velocity.listener.connection.PreConnectListener;
import io.github._4drian3d.authmevelocity.velocity.listener.data.PluginMessageListener;
import io.github._4drian3d.authmevelocity.velocity.listener.input.ChatListener;
import io.github._4drian3d.authmevelocity.velocity.listener.input.CommandListener;
import io.github._4drian3d.authmevelocity.velocity.listener.input.TabCompleteListener;
import java.util.List;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Supplier;
import java.util.stream.Stream;
import net.kyori.adventure.text.logger.slf4j.ComponentLogger;
import org.jetbrains.annotations.NotNull;

public final class AuthMeVelocityPlugin{
  public static final ChannelIdentifier MODERN_CHANNEL
      = MinecraftChannelIdentifier.create("authmevelocity", "main");
  public static final ChannelIdentifier LEGACY_CHANNEL
      = new LegacyChannelIdentifier("authmevelocity:main");

  final Set<String> authServers = ConcurrentHashMap.newKeySet();
  final Set<UUID> loggedPlayers = ConcurrentHashMap.newKeySet();

  private final ProxyServer proxy;
  private final ComponentLogger logger;
  private final Injector injector;

  public AuthMeVelocityPlugin(ProxyServer proxy, ComponentLogger logger, Injector injector) {
    this.proxy = proxy;
    this.logger = logger;
    this.injector = injector;
  }
  
  private MixedLoginConfig config;

  public void onProxyInitialization(MixedLoginConfig config) {
    try {
      authServers.addAll(config.authServers);
    } catch (Exception e) {
      logger.error("Could not load config.conf file", e);
      return;
    }

    logDebug("Loaded plugin libraries");

    proxy.getChannelRegistrar().register(MODERN_CHANNEL, LEGACY_CHANNEL);

    Injector childInjector=injector.createChildInjector(binder -> {
      binder.bind(AuthMeVelocityPlugin.class).toInstance(this);
    });

    Stream.of(
            PluginMessageListener.class,
            DisconnectListener.class,
            InitialServerListener.class,
            PostConnectListener.class,
            PreConnectListener.class,
            ChatListener.class,
            CommandListener.class,
            TabCompleteListener.class
        ).map(childInjector::getInstance)
        .forEach(Listener::register);

    this.sendInfoMessage();
  }

  public void sendInfoMessage() {
    logger.info(miniMessage().deserialize("<gray>AuthServers: <green>" + config.authServers));
    if (config.sendOnLogin.sendOnLogin) {
      logger.info(miniMessage().deserialize(
          "<gray>LobbyServers: <green>" + config.sendOnLogin.teleportServers));
    }
  }

  public void setAuthServers(List<String> servers) {
    authServers.clear();
    authServers.addAll(servers);
  }

  public MixedLoginConfig config() {
    return this.config;
  }

  public boolean isLogged(@NotNull Player player) {
    return loggedPlayers.contains(player.getUniqueId());
  }

  public boolean addPlayer(@NotNull Player player) {
    return loggedPlayers.add(player.getUniqueId());
  }

  public boolean removePlayer(@NotNull Player player) {
    return loggedPlayers.remove(player.getUniqueId());
  }

  public boolean isInAuthServer(@NotNull Player player) {
    return player.getCurrentServer().map(this::isAuthServer).orElse(false);
  }

  public boolean isAuthServer(@NotNull RegisteredServer server) {
    return isAuthServer(server.getServerInfo().getName());
  }

  public boolean isAuthServer(@NotNull ServerConnection connection) {
    return isAuthServer(connection.getServerInfo().getName());
  }

  public boolean isAuthServer(@NotNull String server) {
    return authServers.contains(server);
  }


  public void logDebug(final String msg) {
    if (config.advanced.debug) {
      logger.info("[DEBUG] {}", msg);
    }
  }

  public void logDebug(final Supplier<String> msg) {
    if (config.advanced.debug) {
      logger.info("[DEBUG] {}", msg.get());
    }
  }
}

File: openvc/src/main/kotlin/fun/iiii/mixedlogin/AuthMeVelocityPlugin.kt
package `fun`.iiii.mixedlogin

import com.google.inject.Binder
import com.google.inject.Injector
import com.google.inject.Module
import com.velocitypowered.api.proxy.Player
import com.velocitypowered.api.proxy.ProxyServer
import com.velocitypowered.api.proxy.ServerConnection
import com.velocitypowered.api.proxy.messages.ChannelIdentifier
import com.velocitypowered.api.proxy.messages.LegacyChannelIdentifier
import com.velocitypowered.api.proxy.messages.MinecraftChannelIdentifier
import com.velocitypowered.api.proxy.server.RegisteredServer
import com.velocitypowered.api.util.ModInfo.Mod
import `fun`.iiii.mixedlogin.config.MixedLoginConfig
import `fun`.iiii.mixedlogin.listener.Listener
import `fun`.iiii.mixedlogin.listener.connection.DisconnectListener
import `fun`.iiii.mixedlogin.listener.connection.InitialServerListener
import `fun`.iiii.mixedlogin.listener.connection.PostConnectListener
import `fun`.iiii.mixedlogin.listener.connection.PreConnectListener
import `fun`.iiii.mixedlogin.listener.data.PluginMessageListener
import `fun`.iiii.mixedlogin.listener.input.ChatListener
import `fun`.iiii.mixedlogin.listener.input.CommandListener
import `fun`.iiii.mixedlogin.listener.input.TabCompleteListener
import net.kyori.adventure.text.logger.slf4j.ComponentLogger
import net.kyori.adventure.text.minimessage.MiniMessage
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.function.Supplier
import java.util.stream.Stream

class AuthMeVelocityPlugin(
    private val proxy: ProxyServer,
    private val logger: ComponentLogger,
    private val injector: Injector
) {
    companion object {
        val MODERN_CHANNEL: ChannelIdentifier = MinecraftChannelIdentifier.create("authmevelocity", "main")
        val LEGACY_CHANNEL: ChannelIdentifier = LegacyChannelIdentifier("authmevelocity:main")
    }

    private val authServers: MutableSet<String> = ConcurrentHashMap.newKeySet()
    private val loggedPlayers: MutableSet<UUID> = ConcurrentHashMap.newKeySet()
    private lateinit var config: MixedLoginConfig

    fun onProxyInitialization(config: MixedLoginConfig) {
        try {
            authServers.addAll(config.authServers)
        } catch (e: Exception) {
            logger.error("Could not load config.conf file", e)
            return
        }

        logDebug("Loaded plugin libraries")

        proxy.channelRegistrar.register(MODERN_CHANNEL, LEGACY_CHANNEL)

        val childInjector = injector.createChildInjector(object : Module {
            override fun configure(binder: Binder) {
                binder.bind(AuthMeVelocityPlugin::class.java).toInstance(this@AuthMeVelocityPlugin)
            }
        })


        Stream.of(
            PluginMessageListener::class.java,
            DisconnectListener::class.java,
            InitialServerListener::class.java,
            PostConnectListener::class.java,
            PreConnectListener::class.java,
            ChatListener::class.java,
            CommandListener::class.java,
            TabCompleteListener::class.java
        ).map { clazz: Class<out Listener<*>> -> childInjector.getInstance(clazz) }
            .forEach(Listener<*>::register)

        sendInfoMessage()
    }

    fun sendInfoMessage() {
        logger.info(MiniMessage.miniMessage().deserialize("<gray>AuthServers: <green>$authServers"))
        if (config.sendOnLogin.enable) {
            logger.info(
                MiniMessage.miniMessage().deserialize(
                    "<gray>LobbyServers: <green>${config.sendOnLogin.servers}"
                )
            )
        }
    }

    fun setAuthServers(servers: List<String>) {
        authServers.clear()
        authServers.addAll(servers)
    }

    fun config(): MixedLoginConfig = config

    fun isLogged(player: Player): Boolean = loggedPlayers.contains(player.uniqueId)

    fun addPlayer(player: Player): Boolean = loggedPlayers.add(player.uniqueId)

    fun removePlayer(player: Player): Boolean = loggedPlayers.remove(player.uniqueId)

    fun isInAuthServer(player: Player): Boolean =
        player.currentServer.map { isAuthServer(it) }.orElse(false)

    fun isAuthServer(server: RegisteredServer): Boolean =
        isAuthServer(server.serverInfo.name)

    fun isAuthServer(connection: ServerConnection): Boolean =
        isAuthServer(connection.serverInfo.name)

    fun isAuthServer(server: String): Boolean = authServers.contains(server)

    fun logDebug(msg: String) {
        if (config.advanced.debug) {
            logger.info("[DEBUG] {}", msg)
        }
    }

    fun logDebug(msg: Supplier<String>) {
        if (config.advanced.debug) {
            logger.info("[DEBUG] {}", msg.get())
        }
    }
}